### ISSUE TITLE
Data corrupted when transferring over 4GiB to MPS

### ISSUE THREAD

**Issue by WinnieP** [pytorch/pytorch#124335]
### 🐛 Describe the bug

When converting a cpu tensor to mps, the mps tensor is not the same if the total size is equal to or over 4GiB.

```
import torch

t = torch.ones((2**30-1,), dtype=torch.float32)
t2 = t.to("mps")
print("CPU <4GiB:", (t == torch.tensor(1)).all())
print("MPS <4GiB:", (t2 == torch.tensor(1, device="mps")).all())
print()

t = torch.ones((2**30,), dtype=torch.float32)
t2 = t.to("mps")
print("CPU 4GiB:", (t == torch.tensor(1)).all())
print("MPS 4GiB:", (t2 == torch.tensor(1, device="mps")).all())
```

Output:
```
CPU <4GiB: tensor(True)
MPS <4GiB: tensor(True, device='mps:0')

CPU 4GiB: tensor(True)
MPS 4GiB: tensor(False, device='mps:0')
```

### Versions

PyTorch version: 2.2.2
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.29.0
Libc version: N/A

Python version: 3.11.8 (main, Apr  4 2024, 20:29:28) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.2.2
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @kadeng @mruberry @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by malfet** [comment]
Grabbing for myself to fix the copy

**Comment by malfet** [comment]
This fixes is up to a 6Gb tensors, but underlying problem still remains:
```patch
diff --git a/aten/src/ATen/mps/MPSStream.mm b/aten/src/ATen/mps/MPSStream.mm
index 2ac8b0cc64a..3ab371c3a7b 100644
--- a/aten/src/ATen/mps/MPSStream.mm
+++ b/aten/src/ATen/mps/MPSStream.mm
@@ -173,11 +173,22 @@ void MPSStream::copy(id<MTLBuffer> srcBuffer,
       endKernelCoalescing();
       id<MTLBlitCommandEncoder> blitEncoder = [commandBuffer() blitCommandEncoder];
 
-      [blitEncoder copyFromBuffer:srcBuffer
-                     sourceOffset:(NSUInteger)srcOffset
-                         toBuffer:dstBuffer
-                destinationOffset:(NSUInteger)dstOffset
-                             size:(NSUInteger)length];
+      // For some reason copyFromBuffer for 4Gb fails without returning an error
+      // See https://github.com/pytorch/pytorch/issues/124335
+      // Workaround by batching copy commands into 2Gb chunks
+      constexpr size_t max_copy_size = 0xc0000000; // 3GB
+      size_t bytes_copied = 0;
+      size_t bytes_remains = length;
+      while (bytes_remains > 0) {
+          NSUInteger bytes_to_copy = std::min(max_copy_size, bytes_remains);
+          [blitEncoder copyFromBuffer:srcBuffer
+                         sourceOffset:(NSUInteger)srcOffset + bytes_copied
+                             toBuffer:dstBuffer
+                    destinationOffset:(NSUInteger)dstOffset + bytes_copied
+                                 size:bytes_to_copy];
+          bytes_copied = bytes_to_copy;
+          bytes_remains -= bytes_to_copy;
+      }
       [blitEncoder endEncoding];
 
       // profilerId has a value only if copy profiling is enabled
```

**Comment by malfet** [comment]
Reopening this issue and moving to Kulin as problem still exist in Metal/MPS API

**Comment by kulinseth** [comment]
Agree with @malfet , thanks for opening back this issue. Its indeed high priority..

**Comment by skotapati** [comment]
Confirmed that the issue no longer repros after this fix

**Comment by jhavukainen** [comment]
Closing the issue as the fix has been merged.

**Comment by huydhn** [comment]
This issue has been fixed in the upcoming 2.3.1 release https://dev-discuss.pytorch.org/t/pytorch-release-2-3-1-final-rc-is-available/2126
